internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
+	// Review: our Storage is a bit of a mix of AcquireState and the Ipld{Get-Put}
+	// AcquireState() ActorStateHandle
+
+	// Review: The spec suggest handling return construction by the runtime, interesting but not strictly necessary
+	// SuccessReturn()     msg.InvocOutput
+	// ValueReturn(Bytes)  msg.InvocOutput
+	// ErrorReturn(exitCode exitcode.ExitCode) msg.InvocOutput
+
+	// Review: how does this work? is the spec expecting us to Panic? or just terminate the message execution?
+	// Review: this should be `Abort(string) -> msg.InvocOutput` if we follow the pattern
+	// Abort(string)
+
+	// Review: this is a bit trickier than abort in Go. This is a conditional early exit on the function which we cannot do.
+	// Review: `if <cond> { return AssertFailure("blah") }` is a bit closer to go if we are to try to follow this pattern.
+	// Assert(bool)
```

Comment:
As currently specified, `Assert` just calls `rt.Abort` if the condition is false (`runtime.go:149`).

Reply:
This is very unimplementable without using exceptions or macros. With the `abort()` you can at least return the object, this is a conditional. 

I would also remove it from the interface and go ahead and write `assert(<cond>)` on the pseudocode and just explain the `assert` method magic somewhere else. 

```go
// Assert with abort the execution of the vm call stack if the condition is `False`.
// The caller should receive error 66 `AssertError`. 
func assert(|| cond ||)
```

Implementors can add it back to the interface if they are fine with having it there, but same as with abort, unless the `assert/1` actually needs the `VMContext` object, then it should not be on that interface. Those two seem like free functions/macros.

Reply:
- `rt.Assert` is a function that's callable from VM code, and thus, I think has to be on the `VMContext` (or what we previously called the `Runtime`, i.e., whatever the API object is that is accessible to each actor method). I think it's fine, in languages that make it natural, to use `return rt.AbortReturn(...)` instead of `rt.Abort(...)`, but I'm not convinced that Go is one of those languages (see above).

- We can't write `assert` as a macro in pseudocode, because the intent is for the spec (or at least these parts of it) to be actual Go code -- correct, compilable, and testable (albeit possibly too inefficient to serve as a reference implementation).